core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala

```diff
@@ -0,0 +1,49 @@
+package codesearch.core.index.directory
+
+import java.nio.file.{Path, Paths}
+
+import codesearch.core.index.directory.PathOps._
+import codesearch.core.index.repository._
+
+object PathOps {
+  implicit final class RichPath(val parent: Path) {
+    def /(child: Path): Path   = Paths.get(s"${parent.toFile.getPath}/${child.toFile.getPath}")
+    def /(child: String): Path = Paths.get(s"${parent.toFile.getPath}/$child")
+  }
+}
+
+trait Directory[A <: SourcePackage] {
+  def root: Path        = Paths.get(s"./data")
+  def extension: String = "tgz"
+  def archive(pack: A): Path
+  def unarchived(pack: A): Path = archive(pack).getParent
+}
+
+object PackageDirectory {
+
+  implicit class PackageDirectoryOps[A <: SourcePackage](val pack: A) {
+    def archive(implicit ev: Directory[A]): Path    = ev.archive(pack)
+    def unarchived(implicit ev: Directory[A]): Path = ev.unarchived(pack)
+  }
+
+  implicit def hackageDirectory: Directory[HackagePackage] = new Directory[HackagePackage] {
+    override def archive(pack: HackagePackage): Path =
+      root / "hackage" / s"${pack.name}" / s"${pack.version}" / s"${pack.name}-${pack.version}.$extension"
+  }
+
+  implicit def npmDirectory: Directory[NpmPackage] = new Directory[NpmPackage] {
+    override def archive(pack: NpmPackage): Path =
+      root / "npm" / s"${pack.name}" / s"${pack.version}" / s"${pack.name}-${pack.version}.$extension"
```

Comment:
Will this work with crazy NPM package names?

I don't know how `NpmPackage` is defined but it would make sense to store "true" names there and do the escaping here or in some other place (as long as we don't have five different places where the logic for finding a package is reimplemented).

Maybe we can do that in a separate PR.